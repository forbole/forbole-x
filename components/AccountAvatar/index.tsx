import {
  Box,
  Avatar,
  Typography,
  Snackbar,
  useTheme,
  IconButton,
  TypographyProps,
} from '@material-ui/core';
import React from 'react';
import { Alert } from '@material-ui/lab';
import useTranslation from 'next-translate/useTranslation';
import { Variant } from '@material-ui/core/styles/createTypography';
import CopyIcon from '../../assets/images/icons/icon_copy.svg';
import useIconProps from '../../misc/useIconProps';
import cryptocurrencies from '../../misc/cryptocurrencies';
import useStyles from './styles';
import { CustomTheme } from '../../misc/theme';
import { useWalletsContext } from '../../contexts/WalletsContext';
import LedgerIcon from '../../assets/images/icons/usb_device.svg';

interface AccountAvatarProps {
  address?: {
    address: string;
    crypto: string;
    moniker: string;
    note?: string;
    img?: string;
  };
  account?: Account;
  hideAddress?: boolean;
  disableCopyAddress?: boolean;
  size?: 'small' | 'base' | 'large';
  ledgerIconDisabled?: boolean;
  titleProps?: TypographyProps;
  avatarSize?: number;
}

const AccountAvatar: React.FC<AccountAvatarProps> = ({
  account,
  hideAddress,
  size = 'base',
  disableCopyAddress,
  address,
  ledgerIconDisabled,
  titleProps,
  avatarSize,
}) => {
  const crypto = cryptocurrencies[account ? account.crypto : address.crypto];
  const { t } = useTranslation('common');
  const iconProps = useIconProps();
  const themeStyle: CustomTheme = useTheme();
  const classes = useStyles();
  const [isCopySuccess, setIsCopySuccess] = React.useState(false);
  const { wallets } = useWalletsContext();
  const walletAccountInfo = React.useMemo(
    () =>
      account
        ? { ...account, ...wallets.find(wal => wal.id === account.walletId) }
        : { type: 'mnemonic' },
    [account, wallets],
  );

  let avatarClass = '';
  let titleVariant: Variant = 'h5';
  if (size === 'small') {
    avatarClass = classes.smallAvatar;
    titleVariant = 'body1';
  } else if (size === 'large') {
    avatarClass = classes.largeAvatar;
    titleVariant = 'h3';
  }

  const copyText = React.useCallback(
    e => {
      e.stopPropagation();
      navigator.clipboard.writeText(account ? account.address : address.address);
      setIsCopySuccess(true);
    },
    [account],
  );

  const onClose = React.useCallback(() => {
    setIsCopySuccess(false);
  }, []);

  return (
    <>
      <Box display="flex" alignItems="center">
        <Avatar
          className={avatarClass}
          style={
            avatarSize
              ? { width: themeStyle.spacing(avatarSize), height: themeStyle.spacing(avatarSize) }
              : {}
          }
          alt={crypto.name}
          src={crypto.image}
        />
        <Box mx={1}>
          <Typography variant={titleVariant} {...titleProps}>
            {account ? account.name : address.moniker}
          </Typography>
          {hideAddress || disableCopyAddress ? null : (
            <Box display="flex" alignItems="center" my={-1}>
              <Typography color="textSecondary" variant="body2">
                {account ? account.address : address.address}
              </Typography>
              <IconButton onClick={copyText}>
                <CopyIcon {...iconProps} />
              </IconButton>
            </Box>
          )}
          {disableCopyAddress && !hideAddress ? (
            <Typography variant="body2" color="textSecondary">
              {account ? account.address : address.address}
            </Typography>
          ) : null}
        </Box>
        {walletAccountInfo.type === 'ledger' ? (
          <LedgerIcon
            fill={themeStyle.palette.text.primary}
            style={{ display: ledgerIconDisabled ? 'none' : 'block' }}
          />
        ) : null}
      </Box>
      <Snackbar open={isCopySuccess} autoHideDuration={5000} onClose={onClose}>
        <Alert onClose={onClose} severity="success">
          {t('copied to clipboard')}
        </Alert>
      </Snackbar>
    </>
  );
};

export default AccountAvatar;
