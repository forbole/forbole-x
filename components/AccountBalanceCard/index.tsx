import {
  Avatar,
  Box,
  Card,
  Divider,
  Typography,
  Table,
  TableHead,
  TableBody,
  TableRow,
  TableCell,
  useTheme,
  Button,
} from '@material-ui/core';
import useTranslation from 'next-translate/useTranslation';
import React from 'react';
import get from 'lodash/get';
import { PieChart, Pie, Cell } from 'recharts';
import ArrowNextIcon from '../../assets/images/icons/icon_arrow_next.svg';
import useStyles from './styles';
import { CustomTheme } from '../../misc/theme';
import { formatCrypto, formatCurrency } from '../../misc/utils';
import { useGeneralContext } from '../../contexts/GeneralContext';
import cryptocurrencies from '../../misc/cryptocurrencies';
import useIconProps from '../../misc/useIconProps';
import useIsMobile from '../../misc/useIsMobile';

interface AccountBalanceCardProps {
  account: Account;
  accountBalance: AccountBalance;
  onVestingClick: () => void;
  hideVestingButton?: boolean;
}

const AccountBalanceCard: React.FC<AccountBalanceCardProps> = ({
  accountBalance,
  account,
  onVestingClick,
  hideVestingButton,
}) => {
  const classes = useStyles();
  const { t, lang } = useTranslation('common');
  const theme: CustomTheme = useTheme();
  const { currency, currencyRate, hideAmount } = useGeneralContext();
  const isMobile = useIsMobile();
  const data = Object.keys(accountBalance.balance)
    .filter(
      k =>
        k !== 'commissions' || get(accountBalance, `balance.${k}.${account.crypto}.amount`, 0) > 0,
    )
    .map((k, i) => ({
      name: t(k),
      value: get(accountBalance, `balance.${k}.${account.crypto}.amount`, 0),
      color: theme.palette.pieChart2[i % theme.palette.pieChart2.length],
    }));
  const iconProps = useIconProps();
  const usdPrice = get(accountBalance, `balance.available.${account.crypto}.price`, 0);
  const totalBalance = data.map(d => d.value).reduce((a, b) => a + b, 0);

  // Other Token
  const otherTokens = React.useMemo(() => {
    const balances = {};
    const rewards = {};
    const commissions = {};
    Object.keys(accountBalance.balance).forEach(k => {
      Object.keys(accountBalance.balance[k]).forEach(tk => {
        // Skip native token
        if (tk === account.crypto) {
          return;
        }
        const amount = get(accountBalance, `balance.${k}.${tk}.amount`, 0);
        const price = get(accountBalance, `balance.${k}.${tk}.price`, 0);
        balances[tk] = {
          amount: get(balances, `${tk}.amount`, 0) + amount,
          price,
        };
        if (k === 'rewards') {
          rewards[tk] = {
            amount: get(rewards, `${tk}.amount`, 0) + amount,
            price,
          };
        }
        if (k === 'commissions') {
          commissions[tk] = {
            amount: get(commissions, `${tk}.amount`, 0) + amount,
            price,
          };
        }
      });
    });
    return { balances, rewards, commissions };
  }, [accountBalance, account]);

  return (
    <>
      <Card className={classes.container}>
        <Box display="flex" justifyContent="space-between" alignItems="center">
          <Typography variant="h4">{t('balance')}</Typography>
          {hideVestingButton ? null : (
            <Button onClick={onVestingClick} endIcon={<ArrowNextIcon {...iconProps} />}>
              {t('vesting')}
            </Button>
          )}
        </Box>
        <Box display="flex" my={2} alignItems="center">
          <PieChart
            height={theme.spacing(isMobile ? 16 : 20)}
            width={theme.spacing(isMobile ? 16 : 20)}
          >
            <Pie
              data={data.filter(d => !!d.value)}
              innerRadius={theme.spacing(isMobile ? 7 : 8.5)}
              outerRadius={theme.spacing(isMobile ? 7 : 8.5)}
              paddingAngle={12}
              dataKey="value"
            >
              {data
                .filter(d => !!d.value)
                .map(d => (
                  <Cell
                    key={d.name}
                    strokeLinejoin="round"
                    strokeWidth={theme.spacing(isMobile ? 1.2 : 1.5)}
                    stroke={d.color}
                  />
                ))}
            </Pie>
          </PieChart>
          <Box flex={1} ml={isMobile ? 2 : 10}>
            {data.map(d => (
              <Box key={d.name} display="flex" alignItems="center" justifyContent="space-between">
                <Box display="flex" alignItems="center" mb={0.5}>
                  <Box
                    bgcolor={d.color}
                    width={theme.spacing(1.75)}
                    height={theme.spacing(1.75)}
                    borderRadius={theme.spacing(0.5)}
                    mr={1}
                  />
                  <Typography>{d.name}</Typography>
                </Box>
                <Typography>
                  {formatCrypto(d.value, { unit: account.crypto, lang, hideAmount })}
                </Typography>
              </Box>
            ))}
          </Box>
        </Box>
        <Divider />
        <Box mt={2} display="flex" justifyContent="space-between">
          <Box>
            <Typography variant="h4">{t('total balance')}</Typography>
            <Typography>
              {formatCurrency(usdPrice * currencyRate, {
                currency,
                lang,
                hideUnit: true,
              })}{' '}
              / {account.crypto}
            </Typography>
          </Box>
          <Box display="flex" flexDirection="column" alignItems="flex-end">
            <Typography variant={isMobile ? 'h4' : 'h1'}>
              {formatCrypto(totalBalance, { unit: account.crypto, lang, hideAmount })}
            </Typography>
            <Typography>
              {formatCurrency(usdPrice * totalBalance * currencyRate, {
                currency,
                lang,
                hideAmount,
              })}
            </Typography>
          </Box>
        </Box>
      </Card>
      {Object.keys(otherTokens.balances).length > 0 ? (
        <Card className={classes.container}>
          <Typography variant="h4">{t('other token balance')}</Typography>
          <Table>
            <TableHead>
              <TableRow>
                <TableCell>{t('index')}</TableCell>
                <TableCell>{t('token')}</TableCell>
                <TableCell align="right">{t('rewards')}</TableCell>
                <TableCell align="right">{t('commissions')}</TableCell>
                <TableCell align="right">{t('balance')}</TableCell>
              </TableRow>
            </TableHead>
            <TableBody>
              {Object.keys(otherTokens.balances).map((token, index) => (
                <TableRow key={token} className={classes.tableRow}>
                  <TableCell>#{index + 1}</TableCell>
                  <TableCell>
                    <Box display="flex" alignItems="center">
                      <Avatar
                        className={classes.avatar}
                        src={String(get(cryptocurrencies, `${token}.image`, ''))}
                      />
                      <Typography>{token}</Typography>
                    </Box>
                  </TableCell>
                  <TableCell align="right">
                    {otherTokens.rewards[token]
                      ? formatCrypto(otherTokens.rewards[token].amount, {
                          unit: token,
                          lang,
                          hideAmount,
                        })
                      : ''}
                  </TableCell>
                  <TableCell align="right">
                    {otherTokens.commissions[token]
                      ? formatCrypto(otherTokens.commissions[token].amount, {
                          unit: token,
                          lang,
                          hideAmount,
                        })
                      : ''}
                  </TableCell>
                  <TableCell align="right">
                    {formatCrypto(otherTokens.balances[token].amount, {
                      unit: token,
                      lang,
                      hideAmount,
                    })}
                  </TableCell>
                </TableRow>
              ))}
            </TableBody>
          </Table>
        </Card>
      ) : null}
    </>
  );
};

export default AccountBalanceCard;
