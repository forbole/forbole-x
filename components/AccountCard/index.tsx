import { Box, Card, Avatar, Typography, Button, useTheme } from '@material-ui/core'
import React from 'react'
import { LineChart, Line, YAxis } from 'recharts'
import UpIcon from '@material-ui/icons/ArrowDropUp'
import DownIcon from '@material-ui/icons/ArrowDropDown'
import useTranslation from 'next-translate/useTranslation'
import useStyles from './styles'
import { useSettingsContext } from '../../contexts/SettingsContext'
import { useAccountCardHook } from './hooks'

const AccountCard: React.FC = () => {
  const classes = useStyles()
  const theme = useTheme()
  const { lang } = useTranslation()
  const { currency } = useSettingsContext()
  // TODO: fetch data from backend
  const now = Date.now()
  // const balance = 104387.26485903
  const usdBalance = 626323.54
  const delta = new Array(24).fill(null).map(() => (Math.random() - 0.5) / 10)
  const data = []
  delta.forEach((d, i) => {
    data.unshift({
      time: now - i * 3600000,
      balance: i === 0 ? usdBalance : data[0].balance * (1 + d),
    })
  })
  const lastBalance = data[23].balance
  const firstBalance = data[0].balance
  const diff = Math.abs(lastBalance - firstBalance)
  const percentageChange = Math.round((10000 * diff) / firstBalance) / 100
  const increasing = lastBalance - firstBalance > 0

  // const denom = chainConfig.display.toUpperCase();

  const { accountInfo } = useAccountCardHook('desmos1qpm8wutycha3ncd0u3w9g42v89xnnfs6f9sg8d')
  // console.log('accountInfo', accountInfo)

  return (
    <Card className={classes.container}>
      <Box mb={7} display="flex" alignItems="center" justifyContent="space-between">
        <Box display="flex" alignItems="center">
          <Avatar
            alt="ATOM"
            src="https://research.binance.com/static/images/projects/cosmos-network/cosmoslogo.png"
          />
          <Typography className={classes.name} variant="h5">
            Account 12
          </Typography>
        </Box>
        <Button variant="outlined">Delegate</Button>
      </Box>
      <Box display="flex" alignItems="center" justifyContent="space-between">
        <Box>
          <LineChart width={theme.spacing(20)} height={theme.spacing(5)} data={data}>
            <YAxis domain={['dataMin', 'dataMax']} hide />
            <Line
              type="monotone"
              dataKey="balance"
              stroke={increasing ? theme.palette.success.main : theme.palette.error.main}
              dot={false}
              strokeWidth={2}
            />
          </LineChart>
          <Box display="flex" mt={1} alignItems="center">
            {increasing ? (
              <UpIcon htmlColor={theme.palette.success.main} />
            ) : (
              <DownIcon htmlColor={theme.palette.error.main} />
            )}
            <Box mr={2}>
              <Typography color="textSecondary">{percentageChange}% (24h)</Typography>
            </Box>
            <Typography>
              {new Intl.NumberFormat(lang, {
                style: 'currency',
                currency,
              }).format(diff)}{' '}
              {currency}
            </Typography>
          </Box>
        </Box>
        <Box display="flex" flexDirection="column" alignItems="flex-end">
          <Typography variant="h4" align="right">
            {/* {new Intl.NumberFormat(lang, {
              signDisplay: 'never',
            }).format(accountInfo.total.raw)} */}
            {accountInfo.total.display ? accountInfo.total.display : ''}
          </Typography>
          <Typography variant="h6" align="right">
            {new Intl.NumberFormat(lang, {
              style: 'currency',
              currency,
            }).format(usdBalance)}{' '}
            {currency}
          </Typography>
        </Box>
      </Box>
    </Card>
  )
}

export default AccountCard
