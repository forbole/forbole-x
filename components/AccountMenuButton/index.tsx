import { Button, IconButton, Menu, MenuItem, Box } from '@material-ui/core';
import React from 'react';
import useTranslation from 'next-translate/useTranslation';
import useStyles from './styles';
import MoreIcon from '../../assets/images/icons/icon_more.svg';
import useIconProps from '../../misc/useIconProps';
import ChangeAccountMonikerDialog from './ChangeAccountMonikerDialog';
import DeleteAccountDialog from './DeleteAccountDialog';
import { useWalletsContext } from '../../contexts/WalletsContext';
import ShowAddressOnLedgerDialog from './ShowAddressOnLedgerDialog';

const AccountMenuButton: React.FC<{
  account: Account;
  buttonComponent?: React.ReactElement;
}> = ({ account, buttonComponent }) => {
  const { t } = useTranslation('common');
  const iconProps = useIconProps();
  const classes = useStyles();
  const { wallets } = useWalletsContext();
  const wallet = wallets.find(w => w.id === account.walletId);
  const [anchor, setAnchor] = React.useState<Element>();

  const [changeAccountNameOpen, setChangeAccountNameOpen] = React.useState(false);
  const [deleteAccountOpen, setDeleteAccountOpen] = React.useState(false);
  const [showAddressOnLedgerDialogOpen, setShowAddressOnLedgerDialogOpen] = React.useState(false);

  const onClose = React.useCallback(() => setAnchor(undefined), [setAnchor]);

  return (
    <>
      {buttonComponent ? (
        React.cloneElement(buttonComponent, { onClick: e => setAnchor(e.currentTarget) })
      ) : (
        <IconButton onClick={e => setAnchor(e.currentTarget)}>
          <MoreIcon {...iconProps} />
        </IconButton>
      )}
      <Menu
        anchorEl={anchor}
        getContentAnchorEl={null}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
        keepMounted
        open={!!anchor}
        onClose={onClose}>
        {wallet && wallet.type === 'ledger' ? (
          <MenuItem
            className={classes.menuItem}
            button
            onClick={() => {
              setShowAddressOnLedgerDialogOpen(true);
              onClose();
            }}>
            {t('show address on ledger')}
          </MenuItem>
        ) : null}
        <MenuItem
          className={classes.menuItem}
          button
          onClick={() => {
            setChangeAccountNameOpen(true);
            onClose();
          }}>
          {t('change account moniker')}
        </MenuItem>
        <Box m={2}>
          <Button
            fullWidth
            variant="contained"
            color="primary"
            onClick={() => {
              setDeleteAccountOpen(true);
              onClose();
            }}>
            {t('delete account')}
          </Button>
        </Box>
      </Menu>
      <ChangeAccountMonikerDialog
        open={changeAccountNameOpen}
        onClose={() => setChangeAccountNameOpen(false)}
        account={account}
      />
      <DeleteAccountDialog
        open={deleteAccountOpen}
        onClose={() => setDeleteAccountOpen(false)}
        account={account}
      />
      <ShowAddressOnLedgerDialog
        open={showAddressOnLedgerDialogOpen}
        onClose={() => setShowAddressOnLedgerDialogOpen(false)}
        account={account}
      />
    </>
  );
};

export default React.memo(AccountMenuButton);
