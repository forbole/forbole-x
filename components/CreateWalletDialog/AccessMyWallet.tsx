import {
  Button,
  DialogActions,
  DialogContent,
  Typography,
  Box,
  DialogContentText,
  ButtonBase,
} from '@material-ui/core';
import useTranslation from 'next-translate/useTranslation';
import React from 'react';
import useStyles from './styles';
import { ImportStage } from '.';

interface AccessMyWalletProps {
  onConfirm(stage: ImportStage): void;
  onWhatIsMnemonicClick(): void;
}

const AccessMyWallet: React.FC<AccessMyWalletProps> = ({ onConfirm, onWhatIsMnemonicClick }) => {
  const { t } = useTranslation('common');
  const classes = useStyles();

  return (
    <>
      <DialogContent className={classes.dialogContent}>
        <DialogContentText>{t('access my wallet description')}</DialogContentText>
        <Box my={-2}>
          {Object.values(ImportStage).map(stage => (
            <ButtonBase
              key={stage}
              className={classes.borderedButton}
              onClick={() => onConfirm(stage)}>
              <Typography variant="h6" gutterBottom>
                {t(stage)}
              </Typography>
              <Typography variant="body2" color="textSecondary">
                {t(`${stage} description`)}
              </Typography>
            </ButtonBase>
          ))}
        </Box>
      </DialogContent>
      <DialogActions>
        <Button onClick={onWhatIsMnemonicClick} className={classes.button} color="primary">
          {t('what is secret recovery phrase')}
        </Button>
      </DialogActions>
    </>
  );
};

export default AccessMyWallet;
