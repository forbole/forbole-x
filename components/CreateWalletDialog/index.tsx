import { Dialog, DialogTitle, IconButton } from '@material-ui/core'
import useTranslation from 'next-translate/useTranslation'
import React from 'react'
import { Cosmos } from 'ledger-app-cosmos'
import CloseIcon from '../../assets/images/icons/icon_cross.svg'
import BackIcon from '../../assets/images/icons/icon_back.svg'
import useStyles from './styles'
import useIconProps from '../../misc/useIconProps'
import CreateWallet from './CreateWallet'
import ConfirmMnemonic from './ConfirmMnemonic'
import { useWalletsContext } from '../../contexts/WalletsContext'
import SecurityPassword from './SecurityPassword'
import ImportWallet from './ImportWallet'
import AccessMyWallet from './AccessMyWallet'
import WhatIsMnemonic from './WhatIsMnemonic'
import Start from './Start'
import ImportMnemonicBackup from './ImportMnemonicBackup'
import sendMsgToChromeExt from '../../misc/sendMsgToChromeExt'
import useStateHistory from '../../misc/useStateHistory'
import ConnectLedgerDialogContent from '../ConnectLedgerDialogContent'
import useIsMobile from '../../misc/useIsMobile'
import { connectLedger } from '../../misc/ledger'

let ledgerApp: Cosmos

export enum ImportStage {
  ImportMnemonicPhraseStage = 'import mnemonic phrase',
  MnemonicPhraseBackupStage = 'use mnemonic phrase backup',
  ConnectLedgerDeviceStage = 'connect ledger device',
}

export enum CommonStage {
  StartStage = 'start',
  AccessMyWalletStage = 'access my wallet',
  CreateWalletStage = 'create wallet',
  ConfirmMnemonicStage = 'confirm mnemonic',
  SetSecurityPasswordStage = 'set security password',
  ImportWalletStage = 'import wallet',
  ImportLedgerWalletStage = 'import ledger wallet',
  WhatIsMnemonicStage = 'what is mnemonic',
}

type Stage = CommonStage | ImportStage

interface CreateWalletDialogProps {
  open: boolean
  onClose(): void
  initialStage?: Stage
}

interface Content {
  title: string
  content: React.ReactNode
}

const CreateWalletDialog: React.FC<CreateWalletDialogProps> = ({ open, onClose, initialStage }) => {
  const { t } = useTranslation('common')
  const classes = useStyles()
  const iconProps = useIconProps()
  const { addWallet } = useWalletsContext()
  const isMobile = useIsMobile()
  const [stage, setStage, toPrevStage, isPrevStageAvailable] = useStateHistory<Stage>(
    initialStage || CommonStage.StartStage
  )
  const [mnemonic, setMnemonic] = React.useState('')
  const [securityPassword, setSecurityPassword] = React.useState('')
  const [error, setError] = React.useState('')

  React.useEffect(() => {
    if (open) {
      setMnemonic('')
      setSecurityPassword('')
      setError('')
      setStage(initialStage || CommonStage.StartStage, true)
    }
  }, [open, initialStage])

  const createWallet = React.useCallback(async () => {
    const wallet = await sendMsgToChromeExt({ event: 'generateMnemonic' })
    setMnemonic(wallet.mnemonic)
    setStage(CommonStage.CreateWalletStage)
  }, [setMnemonic])

  const importMnemonic = React.useCallback(
    async (input) => {
      try {
        const wallet = await sendMsgToChromeExt({
          event: 'verifyMnemonic',
          data: { mnemonic: input },
        })
        setMnemonic(wallet.mnemonic)
        setStage(CommonStage.SetSecurityPasswordStage)
      } catch (err) {
        setError(t(err.message))
      }
    },
    [setMnemonic, setStage, setError]
  )

  const importMnemonicBackup = React.useCallback(
    async (data) => {
      try {
        const wallet = await sendMsgToChromeExt({
          event: 'verifyMnemonicBackup',
          data,
        })
        setMnemonic(wallet.mnemonic)
        setStage(CommonStage.SetSecurityPasswordStage)
      } catch (err) {
        setError(t(err.message))
      }
    },
    [setMnemonic, setStage, setError]
  )

  const confirmMnemonic = React.useCallback(
    (input) => {
      if (input === mnemonic) {
        setStage(CommonStage.SetSecurityPasswordStage)
      } else {
        setError(t('invalid mnemonic'))
      }
    },
    [mnemonic, setStage, setError]
  )

  const confirmSecurityPassword = React.useCallback(
    (pw: string) => {
      setSecurityPassword(pw)
      setStage(CommonStage.ImportWalletStage)
    },
    [setStage, setSecurityPassword]
  )

  const saveWallet = React.useCallback(
    async (name: string, cryptos: string[], type = 'mnemonic') => {
      await addWallet(
        {
          type,
          name,
          cryptos,
          mnemonic,
          securityPassword,
        },
        ledgerApp
      )
      onClose()
    },
    [addWallet, mnemonic, securityPassword, onClose]
  )

  const content: Content = React.useMemo(() => {
    switch (stage) {
      case ImportStage.ConnectLedgerDeviceStage:
        return {
          title: '',
          content: (
            <ConnectLedgerDialogContent
              onConnect={async () => {
                ledgerApp = await connectLedger()
                setStage(CommonStage.ImportLedgerWalletStage, undefined, true)
              }}
            />
          ),
        }
      case ImportStage.ImportMnemonicPhraseStage:
        return {
          title: t('mnemonic'),
          content: (
            <ConfirmMnemonic
              description={t('mnemonic description')}
              onConfirm={importMnemonic}
              error={error}
            />
          ),
        }
      case ImportStage.MnemonicPhraseBackupStage:
        return {
          title: t('mnemonic phrase backup title'),
          content: <ImportMnemonicBackup onConfirm={importMnemonicBackup} error={error} />,
        }
      case CommonStage.CreateWalletStage:
        return {
          title: t('create new wallet title'),
          content: (
            <CreateWallet
              mnemonic={mnemonic}
              onConfirm={() => setStage(CommonStage.ConfirmMnemonicStage)}
            />
          ),
        }
      case CommonStage.ConfirmMnemonicStage:
        return {
          title: t('create new wallet title'),
          content: (
            <ConfirmMnemonic
              description={t('confirm mnemonic description')}
              onConfirm={confirmMnemonic}
              error={error}
            />
          ),
        }
      case CommonStage.SetSecurityPasswordStage:
        return {
          title: t('security password title'),
          content: <SecurityPassword onConfirm={confirmSecurityPassword} />,
        }
      case CommonStage.ImportWalletStage:
      case CommonStage.ImportLedgerWalletStage:
        return {
          title: t('import wallet title'),
          content: (
            <ImportWallet
              onConfirm={(name, cryptos) =>
                saveWallet(
                  name,
                  cryptos,
                  stage === CommonStage.ImportLedgerWalletStage ? 'ledger' : 'mnemonic'
                )
              }
            />
          ),
        }
      case CommonStage.WhatIsMnemonicStage:
        return {
          title: t('what is mnemonic phrase'),
          content: <WhatIsMnemonic />,
        }
      case CommonStage.AccessMyWalletStage:
        return {
          title: t('access my wallet title'),
          content: <AccessMyWallet onConfirm={setStage} onCreateWallet={createWallet} />,
        }
      case CommonStage.StartStage:
      default:
        return {
          title: t('create wallet title'),
          content: (
            <Start
              onWhatIsMnemonicClick={() => setStage(CommonStage.WhatIsMnemonicStage)}
              onCreateWalletClick={createWallet}
              onImportWalletClick={() => setStage(CommonStage.AccessMyWalletStage)}
            />
          ),
        }
    }
  }, [stage, t])

  return (
    <Dialog
      fullWidth
      open={open}
      onClose={onClose}
      fullScreen={isMobile}
      PaperProps={{
        className: classes.dialog,
      }}
    >
      {isPrevStageAvailable ? (
        <IconButton className={classes.backButton} onClick={toPrevStage}>
          <BackIcon {...iconProps} />
        </IconButton>
      ) : null}
      <IconButton className={classes.closeButton} onClick={onClose}>
        <CloseIcon {...iconProps} />
      </IconButton>
      <DialogTitle>{content.title}</DialogTitle>
      {content.content}
    </Dialog>
  )
}

export default CreateWalletDialog
