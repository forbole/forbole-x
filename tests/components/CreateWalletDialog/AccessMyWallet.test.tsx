import React from 'react';
import renderer from 'react-test-renderer';
import { ImportStage } from '../../../components/CreateWalletDialog';
import AccessMyWallet from '../../../components/CreateWalletDialog/AccessMyWallet';

const onConfirm = jest.fn();
const onCreateWallet = jest.fn();

describe('component: CreateWalletDialog - AccessMyWallet', () => {
  it('renders correctly', () => {
    const component = renderer.create(
      <AccessMyWallet onConfirm={onConfirm} onCreateWallet={onCreateWallet} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('calls onConfirm with ImportStage.ImportMnemonicPhrase when first button is clicked', () => {
    const component = renderer.create(
      <AccessMyWallet onConfirm={onConfirm} onCreateWallet={onCreateWallet} />,
    );
    renderer.act(() => {
      component.root.findAllByType('button')[0].props.onClick();
    });
    expect(onConfirm).toBeCalledWith(ImportStage.ImportMnemonicPhraseStage);
  });
  it('calls onConfirm with ImportStage.MnemonicPhraseBackup when second button is clicked', () => {
    const component = renderer.create(
      <AccessMyWallet onConfirm={onConfirm} onCreateWallet={onCreateWallet} />,
    );
    renderer.act(() => {
      component.root.findAllByType('button')[1].props.onClick();
    });
    expect(onConfirm).toBeCalledWith(ImportStage.MnemonicPhraseBackupStage);
  });
  it('calls onConfirm with ImportStage.ConnectLedgerDevice when third button is clicked', () => {
    const component = renderer.create(
      <AccessMyWallet onConfirm={onConfirm} onCreateWallet={onCreateWallet} />,
    );
    renderer.act(() => {
      component.root.findAllByType('button')[2].props.onClick();
    });
    expect(onConfirm).toBeCalledWith(ImportStage.ConnectLedgerDeviceStage);
  });
  // it('calls onCreateWallet when last button is clicked', () => {
  //   const component = renderer.create(
  //     <AccessMyWallet onConfirm={onConfirm} onCreateWallet={onCreateWallet} />
  //   )
  //   renderer.act(() => {
  //     component.root.findAllByType('button')[3].props.onClick()
  //   })
  //   expect(onCreateWallet).toBeCalled()
  // })
});

afterEach(() => {
  jest.clearAllMocks();
});
