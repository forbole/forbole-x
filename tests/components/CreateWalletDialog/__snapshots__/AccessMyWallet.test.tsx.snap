// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: CreateWalletDialog - AccessMyWallet renders correctly 1`] = `
Array [
  <div
    className="MuiDialogContent-root HookGlobalStyles-dialogContent-7"
  >
    <p
      className="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      common:access my wallet description
    </p>
    <div
      className="MuiBox-root MuiBox-root-8"
    >
      <button
        className="MuiButtonBase-root HookGlobalStyles-borderedButton-5"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <h6
          className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom"
        >
          common:import mnemonic phrase
        </h6>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          common:import mnemonic phrase description
        </p>
      </button>
      <button
        className="MuiButtonBase-root HookGlobalStyles-borderedButton-5"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <h6
          className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom"
        >
          common:use mnemonic phrase backup
        </h6>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          common:use mnemonic phrase backup description
        </p>
      </button>
      <button
        className="MuiButtonBase-root HookGlobalStyles-borderedButton-5"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <h6
          className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom"
        >
          common:connect ledger device
        </h6>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          common:connect ledger device description
        </p>
      </button>
    </div>
  </div>,
  <div
    className="MuiDialogActions-root MuiDialogActions-spacing"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text HookGlobalStyles-button-1 MuiButton-textSecondary"
      disabled={false}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        common:create a wallet
      </span>
    </button>
  </div>,
]
`;
